hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyRatisContainerReplicaCount.java

```diff
@@ -59,4 +65,45 @@ protected int healthyReplicaCountAdapter() {
   public int getUnhealthyReplicaCountAdapter() {
     return getMisMatchedReplicaCount();
   }
+
+  @Override
+  public boolean isHealthyEnoughForOffline() {
+    long countInService = getReplicas().stream()
+        .filter(r -> r.getDatanodeDetails().getPersistedOpState() == IN_SERVICE)
+        .count();
+    if (countInService == 0) {
+      /*
+      Having no in-service nodes is unexpected and SCM shouldn't allow this
+      to happen in the first place. Return false here just to be safe.
+      */
+      return false;
+    }
+
+    LifeCycleState containerState = getContainer().getState();
+    return (containerState == LifeCycleState.CLOSED
+        || containerState == LifeCycleState.QUASI_CLOSED)
+        && getReplicas().stream()
+        .filter(r -> r.getDatanodeDetails().getPersistedOpState() == IN_SERVICE)
+        .filter(r -> r.getState() !=
+            ContainerReplicaProto.State.UNHEALTHY)
+        .allMatch(r -> ReplicationManager.compareState(
```

Comment:
I'm still not sure about this. What I think method is trying to say is:

Return true if there are some in service nodes that don't have an unhealthy replica?

So if we have:
Container: Closed
IN_SERVICE: Unhealthy
DECOMMISSIONING: Unhealthy

What do we expect here? The method will return true as it stands I think.

But this would return false:

Container: Closed
IN_SERVICE: Quasi_Closed
DECOMMISSIONING: Unhealthy

Reply:
This method is a variation of `isHealthy`, whose purpose is to check if all replicas on in-service nodes are in the same state as the container:
```
  /**
   * Returns true if the container is healthy, meaning all replica which are not
   * in a decommission or maintenance state are in the same state as the
   * container and in QUASI_CLOSED or in CLOSED state.
   *
   * @return true if the container is healthy, false otherwise
   */
  default boolean isHealthy() {
    HddsProtos.LifeCycleState containerState = getContainer().getState();
    return (containerState == HddsProtos.LifeCycleState.CLOSED
        || containerState == HddsProtos.LifeCycleState.QUASI_CLOSED)
        && getReplicas().stream()
        .filter(r -> r.getDatanodeDetails().getPersistedOpState() == IN_SERVICE)
        .allMatch(r -> LegacyReplicationManager.compareState(
            containerState, r.getState()));

  }
```

`isHealthyEnoughForOffline` is intended to do the same, except it shouldn't let `UNHEALTHY` replicas block decommissioning. 

That said, there are situations when we don't want to let UNHEALTHY replicas get decommissioned. For example, when we have a closed/quasi-closed container with only UNHEALTHY replicas. Or when there's a quasi-closed container with UNHEALTHY replicas having the greatest seq id. However this method isn't concerned with that. These cases should be handled by `isSufficientlyReplicatedForOffline`, whose job is to check if we have sufficient number of replicas.


So, I think for this case:
```
Container: Closed
IN_SERVICE: Unhealthy
DECOMMISSIONING: Unhealthy
```
It's ok for this method to return true. However `isSufficientlyReplicatedForOffline` should handle this if we don't have enough healthy replicas.

For:
```
Container: Closed
IN_SERVICE: Quasi_Closed
DECOMMISSIONING: Unhealthy
```
I think the method should return false, because the quasi-closed replica needs to be force closed first.

Reply:
> That said, there are situations when we don't want to let UNHEALTHY replicas get decommissioned. For example, when we have a closed/quasi-closed container with only UNHEALTHY replicas.

What I mean here is that if we have a closed/quasi-container with only UNHEALTHY replicas, and not enough of them. Actually, I was just looking at the decommissioning flow for this situation. I think currently if we have only UNHEALTHY replicas, decommissioning will stay blocked because `isSufficientlyReplicated` doesn't count UNHEALTHY replicas unless the `considerUnhealthy` flag is true. This is the behaviour regardless of this change, and this change doesn't affect it. I think we'll have to look into this situation separately.

Reply:
OK - I think this makes sense as it is then.

> I think currently if we have only UNHEALTHY replicas and enough of them, decommissioning will still stay blocked because isSufficientlyReplicated doesn't count UNHEALTHY replicas unless the considerUnhealthy flag is true. 

For the "new" RM, I think we are going to need to come up with the reconciliation job to fix the unhealthy containers, as its becoming impossible to reason about all these combination of unhealthy states!

Reply:
> For the "new" RM, I think we are going to need to come up with the reconciliation job to fix the unhealthy containers, as its becoming impossible to reason about all these combination of unhealthy states!

I agree!